feat: expose norm1 layer to support TeaCache

andreaferretti opened the pull request:
Fixes https://github.com/mit-han-lab/nunchaku/issues/233

This PR exposes the `norm1` layer from the transformer blocks, which is used by TeaCache. In this way, SVDQuant and TeaCache can be combined to get an even larger speedup on Flux

lmxyy commented:
@ita9naiwa Is TeaCache compatible with First-Block Cache?

ita9naiwa commented:
@lmxyy AFAIK, it's another caching technique and both can't be used at the same time.

@andreaferretti could you show some performance benchmark of this approach?

andreaferretti commented:
Sure, I just made a quick example of inference using:
* A100 machine
* 50 inference steps
* 1024x1024 resolution
* `A cat holding a sign that says hello world` prompt
* guidance_scale 3.5
* seed 0

using Nunchaku 0.1.4 (which I had patched to expose the `norm1` method) both with and without TeaCache. When using TeaCache I have used the default threshold (`r1_rel_tresh`) of 0.6.

Averaging over 10 runs, the Nunchaku version takes 14.9 seconds, while the Nunchaku+TeaCache version takes 4.9 seconds. Results are shown below (first one is without TeaCache, second one is with TeaCache)

The cost of supporting TeaCache is minimal: just make the `norm1` method (or even all the layers) public instead of private. IMO there is no particular reason of keeping the individual layers private, given that other techniques may want to make use of individual layers results, but for TeaCache in particular only `norm1` is needed.

![flux 1-dev-int4](https://github.com/user-attachments/assets/b41669c5-86c6-45c3-81ec-1a1a06c33299)


![flux 1-dev-int4-teacache](https://github.com/user-attachments/assets/b1f50bfd-1d33-43f5-a8c2-5efb6e1bd72e)

andreaferretti commented:
Finally, one question. How does one enable FBCache in Nunchaku? I'd like to compare it to TeaCache

andreaferretti commented:
Thank you! I tried First Block Cache and, while it is working nicely in the basic examples, when I tried to integrate it into more complex pipelines I got results that were either slower or lower quality than TeaCache.

But in any case, I am not arguing to explicitly add support for TeaCache into Nunchaku - just avoid making layers private (at least the norm1 layer) so that the TeaCache script will work as is.

ita9naiwa commented:
thanks for confirming!

to be more clear, could you add some example similar to examples in example/ directory?

I'd like to test too.

andreaferretti commented:
Hi, I added a small example using TeaCache. It uses a context manager that is a slight modification of the official TeaCache script and allows to restore the model state once TeaCache is used, unlike the original script.

ita9naiwa commented on nunchaku/models/transformers/transformer_flux.py:
```diff
@@ -180,6 +180,13 @@ def forward_layer_at(
+        hidden_states: torch.Tensor,
+        emb: torch.Tensor
+    ) -> tuple[torch.Tensor, torch.Tensor, torch.Tensor, torch.Tensor, torch.Tensor]:
+        return self.m.norm1_forward(0, hidden_states, emb)
```
Can we expose not only first layer but arbitrary layer too?
e.g., 
norm1 -> norm_one_at

ita9naiwa replied:
fyi: https://github.com/mit-han-lab/nunchaku/issues/233#issuecomment-2788320686

andreaferretti replied:
We can of course expose every layer, and it would make more sense. But to keep compatibility with the non quantized model, we should keep the name `norm1`.

Regarding the index, this is something that confused me a lot. In the usual Flux model, one get a reference to desired block, and then calls a certain layer on it, something like `model.transformer_blocks[3].norm1`. It is my understanding that in the case of quantized models, one only gets a reference to the collection of all transformer blocks, and then the index of the layer to be called goes into the call, something like `model.transformer_blocks.norm1_at(3)`. Why is the API designed this way? It makes it more difficult to keep compatibility than if one had a reference to the precise block instead

ita9naiwa replied:
https://github.com/mit-han-lab/nunchaku/blob/56728d205f74a41d21f620755beab45772db5d90/nunchaku/models/transformers/transformer_flux.py#L318-L320

We wrap all transformer blocks internally.
(not sure) I guess it is exposed in this way to match compatibility with diffusers

andreaferretti replied:
But you see, `self.transformer_blocks` is a `ModuleList` consisting of a single object. Then whenever you want to do inference on this object, you have to specify an index. This is where the Nunchaku implementation differs from TeaCache, and I am not sure how to reconcile them. The diffusers implementation would have a list of 19 modules in there

ita9naiwa commented on nunchaku/csrc/flux.h:
```diff
@@ -133,6 +133,22 @@ class QuantizedFluxModel : public ModuleWrapper<FluxModel> { // : public torch::
 
+    // expose the norm1 forward method of the transformer blocks
+    // this is used by TeaCache to get the norm1 output
+    std::tuple<torch::Tensor, torch::Tensor, torch::Tensor, torch::Tensor, torch::Tensor> norm1_forward(
```
How about renaming this function to norm_one_forward instead of norm1_forward, even though the actual subprocedure called is norm1.forward, as function names would better contain only alphabets?

andreaferretti replied:
Sure, no problem

andreaferretti replied:
Renamed

ita9naiwa commented:
```
@torch.compiler.disable
def are_two_tensors_similar(t1, t2, *, threshold, parallelized=False):
    mean_diff = (t1 - t2).abs().mean()
    mean_t1 = t1.abs().mean()
    diff = mean_diff / mean_t1
    return diff.item() < threshold
```

I don't understand this math

```python
  coefficients = [
      4.98651651e02,
      -2.83781631e02,
      5.58554382e01,
      -3.82021401e00,
      2.64230861e-01,
  ]
  rescale_func = np.poly1d(coefficients)
  self.accumulated_rel_l1_distance += rescale_func(
      (
          (modulated_inp - self.previous_modulated_input).abs().mean()
          / self.previous_modulated_input.abs().mean()
      )
      .cpu()
      .item()
     )
```

So threshold calculation is bit different but I see their performances matches anyway.

ita9naiwa commented on examples/teacache.py:
```diff
@@ -0,0 +1,400 @@
+logger = logging.get_logger(__name__)  # pylint: disable=invalid-name
+
+
+def make_teacache_forward(
```
Lastly, can we put in `nunchaku/models/transformers/transformer_flux` (btw I don't think it's best place to put this) or somewhere you believe more appropriate?

andreaferretti replied:
I think this should stay as an example. I don't think it makes sense for you to bring TeaCache into your codebase. What does make sense IMO is to expose the layers so that you are API compatible with diffusers, and then any technique that uses these layers will be automatically compatible with Nunchaku

andreaferretti commented:
> I don't understand this math

I did not check it, I just ported the original TeaCache implementation to be compatbile, I am not affiliated with TeaCache in any way

ita9naiwa commented:
Anyway, thanks for the nice work! I checked it runs quite well and shows impressive results!

andreaferretti commented on nunchaku/models/transformers/transformer_flux.py:
```diff
@@ -180,6 +180,13 @@ def forward_layer_at(
+        hidden_states: torch.Tensor,
+        emb: torch.Tensor
+    ) -> tuple[torch.Tensor, torch.Tensor, torch.Tensor, torch.Tensor, torch.Tensor]:
+        return self.m.norm1_forward(0, hidden_states, emb)
```
Done

lmxyy commented:
Could you add a test case in `tests/flux/test_flux_teacache.py` to verify the TeaCache implementation? Please ensure it runs correctly and produces high-quality images without artifacts.

lmxyy commented:
Tests running in https://github.com/mit-han-lab/nunchaku/actions/runs/14634407695/job/41062526118.

lmxyy commented:
> Could you add a test case in `tests/flux/test_flux_teacache.py` to verify the TeaCache implementation? Please ensure it runs correctly and produces high-quality images without artifacts.

Basically, you can first generate several images with the BF16 model, and then with your TeaCache variant. Just need to check the LPIPS of these two sets of images to lower than a specific threshold should be okay.

lmxyy commented:
The tests are passed. Could you add a test case following the guidance in [`tests/README.md`](tests/README.md)?

lmxyy commented:
> Sure, but today is a festivity here, I can add a test case on Monday

Thanks!

andreaferretti commented:
I added a test and rebased on dev. But I had to move the TeaCache implementation out of the example file again, because `flux.1-dev-teacache` was not a valid python module name, and hence I could not import the `TeaCache` class for the test

lmxyy commented:
Running tests in https://github.com/mit-han-lab/nunchaku/actions/runs/14721991356

lmxyy commented on tests/flux/test_flux_teacache.py:
```diff
@@ -0,0 +1,92 @@
+
+    # First, generate results with the 16-bit model
+    if not already_generate(results_dir_16_bit, 1):
+        pipeline = FluxPipeline.from_pretrained(
```
If the GPU does not have enough memory, you need to offload the pipeline to the CPU: [link](https://github.com/mit-han-lab/nunchaku/blob/703bffdebeb7b395f37086aab8d8126babbebb5e/tests/flux/utils.py#L210-L230). Otherwise, the tests may fail on certain servers.

andreaferretti replied:
Should be fixed now

lmxyy reviewed (changes requested):
Just add the adaptive CPU offloading in BF16 tests.

lmxyy commented:
test running in https://github.com/mit-han-lab/nunchaku/actions/runs/14737314849

lmxyy commented on tests/flux/test_flux_teacache.py:
```diff
@@ -0,0 +1,144 @@
+    if not already_generate(results_dir_16_bit, 1):
+        pipeline = FluxPipeline.from_pretrained(
+            "black-forest-labs/FLUX.1-dev", torch_dtype=torch.bfloat16
+        ).to(device)
```
remove this `.to(device)`. Otherwise, the offloading has no use.

andreaferretti replied:
Done

andreaferretti commented:
Please, notice that I will be off for a couple weeks starting tomorrow, so I will not be able to make further adjustments. In case there are still minor adjustments to make to the PR, feel free to edit it directly

lmxyy commented:
I've made some changes: I moved `teacache.py` from the examples directory into the main `nunchaku` library. Also, please avoid overriding the forward method using `self.model.__class__.forward`, as this can interfere with the `__del__` method and lead to unexpected behavior.

lmxyy commented:
tests running in https://github.com/mit-han-lab/nunchaku/actions/runs/14768987873/job/41465923852

lmxyy commented:
@andreaferretti Thanks for your PR to Nunchaku! Would you be interested in contributing more? We have a lot of features planned and would love to collaborate with you. Let me know!

andreaferretti commented:
Hello @lmxyy , sorry for the late response. While I may occasionally provide more PRs, I am doing so in a work context, so it all depends on work priorities that I cannot easily predict at the moment. Thank you again for this great library!
